2009-10-14  Michael Tautschnig  <mt@debian.org>

	* fai-vol_id: New wrapper around udev's vol_id or blkid, whichever is
		available
	* setup-storage/Fstab.pm: Use fai-vol_id
Index: trunk/lib/fai-vol_id
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ trunk/lib/fai-vol_id	
@@ -0,0 +1,48 @@
+#! /bin/bash
+
+# Copyright (c) 2009 by Michael Tautschnig <mt@debian.org>
+
+udev_vol_id="/lib/udev/vol_id"
+udev_blkid="/sbin/blkid"
+
+get_uuid() {
+  if [ -z $udev_vol_id ] ; then
+    $udev_blkid -s UUID -o value $1
+    exit $?
+  fi
+
+  /lib/udev/vol_id -u $1
+  exit $?
+}
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+get_label() {
+  if [ -z $udev_vol_id ] ; then
+    $udev_blkid -s LABEL -o value $1
+    exit $?
+  fi
+
+  /lib/udev/vol_id -l $1
+  exitcode=$?
+  if [ $exitcode -eq 0 -o $exitcode -eq 3 ] ; then
+    exitcode=0
+  fi
+  exit $exitcode
+}
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+# main program
+
+if [ ! -e $udev_vol_id ] ; then
+  udev_vol_id=""
+  if [ ! -e $udev_blkid ] ; then
+    "Neither udev vol_id nor blkid found!"
+    exit 1
+  fi
+fi
+
+while getopts ul opt ; do
+      case "$opt" in
+        u) shift ; get_uuid $1 ;;
+        l) shift ; get_label $1 ;;
+      esac
+done
Index: trunk/lib/setup-storage/Fstab.pm
===================================================================
--- trunk.orig/lib/setup-storage/Fstab.pm
+++ trunk/lib/setup-storage/Fstab.pm	
@@ -94,7 +94,7 @@
   # or labels, use these if available
   my @uuid = ();
   &FAI::execute_ro_command(
-    "/lib/udev/vol_id -u $device_name", \@uuid, 0);
+    "/usr/lib/fai/fai-vol_id -u $device_name", \@uuid, 0);
 
   # every device must have a uuid, otherwise this is an error (unless we
   # are testing only)
@@ -107,8 +107,7 @@
   # ok here
   my @label = ();
   &FAI::execute_ro_command(
-    "( /lib/udev/vol_id -l $device_name ; exc=\$? ; if [ \$exc -eq 3 ] ;" .
-    " then exit 0 ; else exit \$exc ; fi )", \@label, 0);
+    "/usr/lib/fai/fai-vol_id -l $device_name", \@label, 0);
 
   # using the fstabkey value the desired device entry is defined
   if ($key_type eq "uuid") {
Index: trunk/debian/fai-client.install
===================================================================
--- trunk.orig/debian/fai-client.install
+++ trunk/debian/fai-client.install	
@@ -10,6 +10,7 @@
 usr/lib/fai/prcopyleft
 usr/lib/fai/get-config-dir*
 usr/lib/fai/fai-divert
+usr/lib/fai/fai-vol_id
 usr/bin/device2grub
 usr/bin/fai-class
 usr/bin/fai-debconf
